crates/telio-proto/src/packet/relayed/upgrade.rs

```diff
@@ -71,17 +76,125 @@ impl DowncastPacket<PacketRelayed> for UpgradeMsg {
     }
 }
 
+/// Decision of the other node if the upgrade request was accepted and performed
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Decision {
+    /// Upgrade request was accepted and performed
+    Accepted = 0,
+    /// Upgrade request was rejected because it contained unknown session id
+    RejectedDueToUnknownSession = 1,
+    /// Upgrade request was rejected because it was sent when the other node has
+    /// already also sent an upgrade request to us, and they have a winning pub key.
+    RejectedDueToConcurrentUpgrade = 2,
+}
+
+impl From<crate::messages::upgrade::Decision> for Decision {
+    fn from(value: crate::messages::upgrade::Decision) -> Self {
+        match value {
+            crate::messages::upgrade::Decision::Accepted => Decision::Accepted,
+            crate::messages::upgrade::Decision::RejectedDueToUnknownSession => {
+                Decision::RejectedDueToUnknownSession
+            }
+            crate::messages::upgrade::Decision::RejectedDueToConcurrentUpgrade => {
+                Decision::RejectedDueToConcurrentUpgrade
+            }
+        }
+    }
+}
+
+impl From<Decision> for crate::messages::upgrade::Decision {
+    fn from(value: Decision) -> Self {
+        use crate::messages::upgrade::Decision::*;
+        match value {
+            Decision::Accepted => Accepted,
+            Decision::RejectedDueToUnknownSession => RejectedDueToUnknownSession,
+            Decision::RejectedDueToConcurrentUpgrade => RejectedDueToConcurrentUpgrade,
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+/// Result of previously sent upgrade request
+pub struct UpgradeDecisionMsg {
+    /// Decision if the upgrade was accepted and processed
+    pub decision: Decision,
+    /// Session identifing request for which this result in an answer
+    pub session: Session,
+}
+
+impl Codec<PacketTypeRelayed> for UpgradeDecisionMsg {
+    const TYPES: &'static [PacketTypeRelayed] = &[PacketTypeRelayed::UpgradeDecision];
+
+    fn decode(bytes: &[u8]) -> CodecResult<Self>
+    where
+        Self: Sized,
+    {
+        if bytes.is_empty() {
+            return Err(CodecError::InvalidLength);
+        }
+
+        match PacketTypeRelayed::from(*bytes.first().unwrap_or(&(PacketTypeRelayed::Invalid as u8)))
+        {
+            PacketTypeRelayed::UpgradeDecision => {
+                let proto_upgrade_decision = UpgradeDecision::parse_from_bytes(
+                    bytes.get(1..).ok_or(CodecError::DecodeFailed)?,
```

Comment:
You've already checked that bytes len is more than 1 twice. IMHO `bytes[1..]` is OK for this one

Reply:
I think I will combine both of your suggestions into `split_first`.